contrib/auto-render/splitAtDelimiters.js

```diff
@@ -27,76 +27,72 @@ const findEndOfMath = function(delimiter, text, startIndex) {
-            let nextIndex;
+    let lookingForLeft = true;
+    let index;
+    const data = [];
 
-            nextIndex = text.indexOf(leftDelim);
-            if (nextIndex !== -1) {
-                currIndex = nextIndex;
-                finalData.push({
+    while (true) {
+        if (lookingForLeft) {
+            index = text.search(regexLeft);
+            if (index === -1) {
+                break;
+            }
+            if (index > 0) {
+                data.push({
                     type: "text",
-                    data: text.slice(0, currIndex),
+                    data: text.slice(0, index),
                 });
-                lookingForLeft = false;
+                text = text.slice(index);
             }
-
-            while (true) {
-                if (lookingForLeft) {
-                    nextIndex = text.indexOf(leftDelim, currIndex);
-                    if (nextIndex === -1) {
-                        break;
-                    }
-
-                    finalData.push({
-                        type: "text",
-                        data: text.slice(currIndex, nextIndex),
-                    });
-
-                    currIndex = nextIndex;
-                } else {
-                    nextIndex = findEndOfMath(
-                        rightDelim,
-                        text,
-                        currIndex + leftDelim.length);
-                    if (nextIndex === -1) {
-                        break;
-                    }
-
-                    finalData.push({
-                        type: "math",
-                        data: text.slice(
-                            currIndex + leftDelim.length,
-                            nextIndex),
-                        rawData: text.slice(
-                            currIndex,
-                            nextIndex + rightDelim.length),
-                        display: display,
-                    });
-
-                    currIndex = nextIndex + rightDelim.length;
-                }
-
-                lookingForLeft = !lookingForLeft;
+        } else {
+            let i = 0;
+            while (!text.startsWith(delimiters[i].left)) { i++; }
+            index = findEndOfMath(
+                delimiters[i].right,
+                text,
+                delimiters[i].left.length
+            );
+            if (index === -1) {
+                break;
             }
 
-            finalData.push({
-                type: "text",
-                data: text.slice(currIndex),
+            data.push({
+                type: "math",
+                data: text.slice(delimiters[i].left.length, index),
+                rawData: text.slice(0, index + delimiters[i].right.length),
+                display: delimiters[i].display || false,
             });
-        } else {
-            finalData.push(startData[i]);
+            text = text.slice(index + delimiters[i].right.length);
         }
+
+        lookingForLeft = !lookingForLeft;
```

Comment:
I think it'd be a bit easier to grok if the `if` and `else` blocks above set this explicitly, e.g.
```
if (lookingForLeft) {
   ...
   lookingForLeft = false;
} else {
   ...
   lookingForLeft = true;
}
```

Reply:
The more I look at this I wonder if the `if` and `else` blocks are even necessary.  We start out with `lookingForLeft` set to `true` and once we found the "left" we always switch to finding the "right".  Maybe we can get rid of `lookingForLeft` and the `if`/`else` blocks altogether.  That would also mean that we could get rid of some of the variables defined before the `while` loop.

Reply:
sure, I just kept the existing code.